He over wifi

atanasdb opened the pull request.

Commit:
Allow HE to be tried based on mcc/mnc information

Jira: MC-1556

HE service is being extended to support Force HE over WiFi. For the
purpose we need to support HE flows started based on the MCC/MNC
passed by the device to the authorization endpoint.

See design doc at
https://confluence.skunk-works.no/confluence/x/30BRAg.

Commit:
Add HE IPs parsing in the well known response

Commit:
Get WiFi and cellular interfaces

Commit:
Intercept HEI calls

gorandostanic commented on connect/src/com/telenor/connect/ConnectSdk.java:
```diff
@@ -45,6 +52,9 @@
     private static String sPaymentSuccessUri;
     private static SdkProfile sdkProfile;
+    private static ConnectivityManager connectivityManager;
+    private static Network cellularNetwork;
```
Just a control question, since these are initalised only once... are you sure they do not need any reinitialisation if the user changes something on the device (turns off WiFi, ejects SIM card, inserts another SIM or similar)?

atanasdb replied:
Yes, good question.
I haven't managed to test that yet on a real device, as I've only tested on the emulator.
I'm waiting for the final patch to snowball, and then I'll do more thorough tests on a real device and will test what you suggested here.

atanasdb replied:
I wrote a very long description of why this is safe at it is (in case of turning on/off the interfaces). Whoever is interested might read it.
It's in the design code doc, where the 
public WebResourceResponse fetchUrlTroughCellular(String originalUrl) is described.
The doc is at: https://confluence.skunk-works.no/confluence/display/Scale/Facilitating+Header+Enrichment+over+WiFi+in+MobileConnect+%28and+Connect+ID%29+SDK

I decided not to put those comments in the repo as they might expose some TelenorNorway operations stuff.

jorunfa commented on connect/src/com/telenor/connect/ConnectSdk.java:
```diff
@@ -339,6 +352,22 @@ public static synchronized void sdkInitialize(Context context) {
+        return tel.getNetworkOperator();
+    }
+
+    public static String getMccMncLoginHint() {
```
Since this is the main entry file, I think it would make sense to write out these abbreviations, maybe?

atanasdb replied:
Could you pls clarify? MCC MNC are quite standard telco words.
I moved the getMccMncLoginHint though to ConnectLoginButton as it seems this method has much more knowledge for the OAuth query param names.

jorunfa commented on connect/src/com/telenor/connect/ConnectSdk.java:
```diff
@@ -339,6 +352,22 @@ public static synchronized void sdkInitialize(Context context) {
+
+    public static String getMccMncLoginHint() {
+        String mccMnc = getMccMnc();
+        if (mccMnc.isEmpty()) {
```
Change to `return mccMnc.isEmpty() ? null : String.format("MCCMNC:%s", mccMnc);'?

atanasdb replied:
Refactored a bit so no need for that.

jorunfa commented on connect/src/com/telenor/connect/ConnectSdk.java:
```diff
@@ -478,6 +507,8 @@ public static synchronized void sdkInitializeMobileConnect(
                 SimCardStateChangedBroadcastReceiver.getReceiver(),
                 SimCardStateChangedBroadcastReceiver.getIntentFilter());
+
+        initializeCommonComponents();
```
Why is this both here and in `sdkInitialize` method?

atanasdb replied:
I think Goran's plan is for the app developer to either init the connectSDK (sdkInitialize) or MC SDK (sdkInitializeMobileConnect). initializeCommonComponents handles common inits for both, thus in a separate method called by the two.

jorunfa commented on connect/src/com/telenor/connect/ConnectSdk.java:
```diff
@@ -487,4 +518,67 @@ public static synchronized void sdkReinitializeMobileConnect() {
+
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    public static void initalizeCellularNetwork() {
+        NetworkRequest.Builder builder = new NetworkRequest.Builder();
```
think you can chain together the Builder statements.
```java
NetworkRequest networkRequest = NetworkRequest.Builder()
  . addCapability()
  . addTransportType()
  .build()
```

atanasdb replied:
Done

jorunfa commented on connect/src/com/telenor/connect/ConnectSdk.java:
```diff
@@ -487,4 +518,67 @@ public static synchronized void sdkReinitializeMobileConnect() {
+
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    public static void initalizeWiFiNetwork() {
+        NetworkRequest.Builder builder = new NetworkRequest.Builder();
```
Same as above.

atanasdb replied:
Done

jorunfa commented on connect/src/com/telenor/connect/ConnectSdk.java:
```diff
@@ -487,4 +518,67 @@ public static synchronized void sdkReinitializeMobileConnect() {
+            return;
+        }
+        initalizeCellularNetwork();
+        initalizeWiFiNetwork();
```
`@TargetApi` is just lint suppression, it doesn't hinder the code from being run on older versions:
> Indicates that Lint should treat this type as targeting a given API level, no matter what the project target is.

So this code will currently be run no matter what the api level is, it will just crash on older versions. You need to check that and handle them accordingly.

```java
if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
  initalizeWiFiNetwork();
else {
  handleOlderVersion();
}
```

atanasdb replied:
Oh right ... sure. Done

jorunfa commented on connect/src/com/telenor/connect/ui/ConnectLoginButton.java:
```diff
@@ -107,6 +107,11 @@ public void onClick(View v) {
                 parameters.put("scope", TextUtils.join(" ", getLoginScopeTokens()));
             }
 
+            final String mccMncLoginHint = ConnectSdk.getMccMncLoginHint();
```
What is a `MccMncLoginHint`?

atanasdb replied:
For more context I guess you can see the design doc in Confluence? But in short it's the MCCMNC pair sent as a login_hint, so that HE service could pick the right BU data reader from it.

jorunfa replied:
Remember this is an open repo, so Confluence page won't do. But I googled `MCCMNC` and then I understood, so I guess that might be enough for it to make sense for most people.

jorunfa commented on connect/src/com/telenor/connect/ui/ConnectWebViewClient.java:
```diff
@@ -94,6 +109,70 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
     @Override
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
+        // Only implement interception logic when we have both cellular and WiFi interfaces active
```
Remove comment and change to `boolean shouldActivateInterceptionLogic = ConnectSdk.getCellularNetwork() == null || ConnectSdk.getWifiNetwork() == null`

atanasdb replied:
Done

jorunfa commented on connect/src/com/telenor/connect/ui/ConnectWebViewClient.java:
```diff
@@ -94,6 +109,70 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
+    }
+
+    public boolean shouldFetchThroughCellular(String url) {
+        try {
```
Can all of these methods inside the try throw? If not you should move the ones you can out of the try statement.

atanasdb replied:
Yes. All except the one in the return clause. I think it's more readable this way.

jorunfa commented on connect/src/com/telenor/connect/ui/ConnectWebViewClient.java:
```diff
@@ -94,6 +109,70 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
+            return null;
+        }
+        if (shouldFetchThroughCellular(request.getUrl().toString())) {
+            return fetchUrlThroghCellular(request.getUrl().toString());
```
Same as previously. You need to check version code and handle older versions, that is, probably don't call this method if the device is older than Lollipop.

atanasdb replied:
I believe it's safe here as it's overridden and I guess the WebView will simply not call it in case of older versions. What do you think?

jorunfa commented on connect/src/com/telenor/connect/ui/ConnectWebViewClient.java:
```diff
@@ -94,6 +109,70 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
+            } catch (final IOException e) {
+                return null;
+            }
+        } while (true);
```
Change to

```java
} while (attempts <= ConnectSdk.MAX_REDIRECTS_TO_FOLLOW_FOR_HE);
return null;
```
for clarity?

atanasdb replied:
Done

jorunfa commented on connect/src/com/telenor/connect/ui/ConnectWebViewClient.java:
```diff
@@ -94,6 +109,70 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
+        int attempts = 0;
+        do {
+            Network interfaceToUse = ConnectSdk.getCellularNetwork();
+            HttpURLConnection connection = null;
```
`= null;` is redundant. Just use `HttpURLConnection connection;`?

atanasdb replied:
Done

jorunfa commented on connect/src/com/telenor/connect/ui/ConnectWebViewClient.java:
```diff
@@ -94,6 +109,70 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
+            Network interfaceToUse = ConnectSdk.getCellularNetwork();
+            HttpURLConnection connection = null;
+            InputStream inputStream = null;
+            try {
```
Do all these statements need to be inside a try closure?

atanasdb replied:
Yes. All of them except getContentType & getContentEncoding throw so I think it's more readable.

jorunfa reviewed (changes requested):
I very much like the idea of automatic network switching for HE 👍

Commit:
Addressing PR comments

Commit:
Using the proper interface after redirect

Commit:
Fix possible NullPointerException

gorandostanic commented on connect/src/com/telenor/connect/ui/ConnectWebViewClient.java:
```diff
@@ -94,6 +109,75 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
+        }
+        return ConnectSdk
+                .getWellKnownConfig()
+                .getNetworkAuthenticationTargetIps()
```
Just a control question... I do not see that you tackle the situation when wellKnowConfig() returns null, let alone getNetworkAuthenticationTargetIps() returning null. For operators other than us, it can happen, can't it?

atanasdb replied:
I'm not sure that I understand. 
Isn't that what this block in the beginning is doing?

gorandostanic replied:
Ah, sorry, my bad (bloody GitHub :-)

Commit:
Using the proper interface for TnN redirects

Commit:
Declaring Network iface references as volatile

Network iface references might be acceessed both by ConnectivityManager
launched thread and by the WebView.

Commit:
Merge branch 'MC-1568-OD' into he_over_wifi

Commit:
Follow the TnN https redirects with cellular interface

jorunfa commented on connect/src/com/telenor/connect/ConnectSdk.java:
```diff
@@ -487,4 +510,65 @@ public static synchronized void sdkReinitializeMobileConnect() {
+            return;
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            initalizeCellularNetwork();
```
Nit picking, but could you reverse this if, make it return from the method if it is true, and then put it at the top of the method (guard)? We don't need the `ConnectivityManager` if we're on an old version.

atanasdb replied:
Done

jorunfa reviewed (approved):
LGTM with one nitpick

Commit:
Refactoring

Commit:
Merge branch 'MC-1568-OD' into he_over_wifi

Commit:
Merge branch 'MC-1568-OD' into he_over_wifi
